crates/matrix-sdk-ui/src/room_list_service/room_list.rs

```diff
@@ -190,31 +198,71 @@ pub enum RoomListLoadingState {
 
 type BoxedFilterFn = Box<dyn Fn(&RoomListEntry) -> bool + Send + Sync>;
 
-/// Dynamic filter for the [`RoomList`] entries.
+/// Controller for the [`RoomList`] dynamic entries.
 ///
-/// To get one value of this type, use [`RoomList::entries_with_dynamic_filter`]
-pub struct DynamicRoomListFilter {
-    inner: Arc<AsyncCell<BoxedFilterFn>>,
+/// To get one value of this type, use
+/// [`RoomList::entries_with_dynamic_adapters`]
+pub struct RoomListDynamicEntriesController {
+    filter: Arc<AsyncCell<BoxedFilterFn>>,
+    page_size: usize,
+    limit: SharedObservable<usize>,
+    maximum_number_of_rooms: Mutex<Subscriber<Option<u32>>>,
 }
 
-impl DynamicRoomListFilter {
-    fn new(inner: Arc<AsyncCell<BoxedFilterFn>>) -> Self {
-        Self { inner }
+impl RoomListDynamicEntriesController {
+    fn new(
+        filter: Arc<AsyncCell<BoxedFilterFn>>,
+        page_size: usize,
+        limit_stream: SharedObservable<usize>,
+        maximum_number_of_rooms: Subscriber<Option<u32>>,
+    ) -> Self {
+        Self {
+            filter,
+            page_size,
+            limit: limit_stream,
+            maximum_number_of_rooms: Mutex::new(maximum_number_of_rooms),
+        }
     }
 
     /// Set the filter.
     ///
     /// If the associated stream has been dropped, returns `false` to indicate
     /// the operation didn't have an effect.
-    pub fn set(&self, filter: impl Fn(&RoomListEntry) -> bool + Send + Sync + 'static) -> bool {
-        if Arc::strong_count(&self.inner) == 1 {
+    pub fn set_filter(
+        &self,
+        filter: impl Fn(&RoomListEntry) -> bool + Send + Sync + 'static,
+    ) -> bool {
+        if Arc::strong_count(&self.filter) == 1 {
             // there is no other reference to the boxed filter fn, setting it
             // would be pointless (no new references can be created from self,
             // either)
             false
         } else {
-            self.inner.set(Box::new(filter));
+            self.filter.set(Box::new(filter));
             true
         }
     }
+
+    /// Add one page, i.e. view `page_size` more entries in the room list if
+    /// any.
+    pub fn add_one_page(&self) {
+        if let Some(max) = self.maximum_number_of_rooms.lock().unwrap().next_now() {
+            let max: usize = max.try_into().unwrap();
+            let limit = self.limit.get();
+
+            if limit < max {
+                // With this logic, it is possible that `limit` becomes greater than `max` if
+                // `max - limit < page_size`, and that's perfectly fine. It's OK to have a
+                // `limit` greater than `max`, but it's not OK to increase the limit
+                // indefinitely.
+                self.limit.set(limit + self.page_size);
```

Comment:
Should we do something like `self.limit.set(max.min(limit + self.page_size))` so we don't even need the comment?

Reply:
Nop, we don't want to clamp or max the result. That could be surprising, and that's why I left the comment :-).

Reply:
I'm not maxing, but taking the minimum of `max` and `limit + self.page_size`?